x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java

```diff
@@ -101,4 +100,54 @@ public BlockFactory blockFactory() {
     public boolean isReleased() {
         return released;
     }
+
+    @Override
+    public final void incRef() {
+        if (references <= 0) {
+            throw new IllegalStateException("can't increase refCount on already released block [" + this + "]");
+        }
+        references++;
+    }
+
+    @Override
+    public final boolean tryIncRef() {
+        if (references <= 0) {
+            return false;
+        }
+        references++;
+        return true;
+    }
+
+    @Override
+    public final boolean decRef() {
+        if (references <= 0) {
+            throw new IllegalStateException("can't release already released block [" + this + "]");
+        }
+
+        references--;
+
+        if (references <= 0) {
+            close();
```

Comment:
The consistent pattern is that the `close` method calls the `decRef` method. In turn, the `decRef` method invokes either `closeInternal` or `doClose` when the references reach zero, releasing the memory. This ensures that subclasses can only override `closeInternal` or `doClose`, not the `close` method.